towhee/compiler/utils/callstack.py

```
@@ -11,55 +11,128 @@
+        and `end` (includes `start`, excludes `end`).
 
         Args:
             start: the index of the start frame.
             end: the index of the end frame.
-            items: the items to be hashed. Supported items are {filename, lineno, 
-                funcname, codectx, index, lasti}, where codectx denotes the current
-                line of code of the context, index denotes the frame's index of the
-                callstack, lasti denotes the index of last attempted instruction in 
-                bytecode.
+            items: the items to be hashed. Supported items are {filename, lineno,
+                function, code_context, position, lasti}, where code_context denotes 
+                the current line of code of the context, position denotes the frame's 
+                index of the callstack, lasti denotes the index of last attempted 
+                instruction in bytecode.
 
         Returns:
             The hash value.
 
         Raises:
+            AttributeError: If an item in `items` is not supported, i.e. not one of
+            {filename, lineno, function, code_context, position, lasti}.
 
         Examples:
+            Suppose we have a callstack obejct `s` contains 2 frames.
+            >>> hash_val = s.hash(0,1,['filename'])
+            >>> print(hash_val)
+            b284a28710cce90d9d9be3a7f4cabc8e 
+            >>> hash_val = s.hash(0,5,['filename'])
+            >>> print(hash_val)
+            ERROR:root:index range [0, 4] out of list range [0, 1]
+            None 
+            >>> hash_val = s.hash(0,1,['attr_a'])
+            >>> print(hash_val)
+            ERROR:root:{'attr_a'} not supported
+            None
         """
-        raise NotImplementedError
+        start = start or 0
+        end = end or self.stack_size
+
+        if end > self.stack_size or end <= 0 or start >= self.stack_size or start < 0 :
+            logging.error(
```

raise ValueError with the msg, rather than logging it